runtime/src/main/scala/com/armanbilge/vilcacora/runtime/Interpreter.scala

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright 2023 Arman Bilge
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.armanbilge.vilcacora.runtime
+
+import cats.effect.{IO, Resource}
+import cats.syntax.all._
+import com.armanbilge.vilcacora.ir._
+import com.armanbilge.vilcacora.runtime.LibSvm._
+import scala.scalanative.unsafe._
+import scala.scalanative.libc.stdlib
+import scala.scalanative.libc.string.memcpy
+import scala.scalanative.unsigned._
+
+/** The core execution engine for a translated `ModelIR`. It manages memory and executes operations
+  * within a Cats Effect IO context.
+  */
+object Interpreter {
+
+  // A type alias for clarity: maps tensor names to their C memory pointers.
+  type MemoryMap = Map[String, Ptr[Byte]]
+
+  /** Executes a complete ModelIR graph from inputs to outputs. */
+  def execute(model: ModelIR, inputs: Map[String, Array[_]]): IO[Map[String, Array[_]]] =
+    // `Resource.use` guarantees all tensor memory is pre-allocated and safely freed.
+    memoryResource(model).use { memoryMap =>
+      for {
+        _ <- copyInputsToMemory(inputs, model, memoryMap)
+
+        // The core execution loop. It executes each operation and then yields control
+        // to the Cats Effect runtime to maintain application responsiveness.
+        // This is the desired structure.
+        _ <- model.operations.traverse_ { op =>
+          executeOperation(op, memoryMap, model) <* IO.cede
+        }
+
+        results <- copyOutputsFromMemory(model, memoryMap)
+      } yield results
```

Comment:
`for ... yield` is syntax sugar for `flatMap`, so instead we should write with `*>` (`productR`) so that we allocate once, up-front.